--- a/fs/squashfs/squashfs_fs.h	2014-07-18 06:58:15.000000000 +0800
+++ b/fs/squashfs/squashfs_fs.h	2014-08-02 19:58:39.794438634 +0800
@@ -240,12 +240,12 @@
 
 struct squashfs_super_block {
 	__le32			s_magic;
-	__le32			inodes;
 	__le32			mkfs_time;
-	__le32			block_size;
+	__le32			inodes;
 	__le32			fragments;
-	__le16			compression;
+	__le32			block_size;
 	__le16			block_log;
+	__le16			compression;
 	__le16			flags;
 	__le16			no_ids;
 	__le16			s_major;
