--- a/source3/wscript_build
+++ b/source3/wscript_build
@@ -1165,7 +1165,7 @@
                  secrets3
                  param''')
 
-smbstatus_source = 'utils/status.c smbd/notify_internal.c'
+smbstatus_source = 'utils/status.c'
 
 if bld.CONFIG_GET("WITH_PROFILE"):
     smbstatus_source += ' utils/status_profile.c'
@@ -1469,6 +1469,68 @@
                   deps='ncurses menu panel form registry param popt_samba3 smbregistry',
                   enabled=bld.env.build_regedit)
 
+
+bld.SAMBA3_SUBSYSTEM('SMBD_MULTI',
+                     source='smbd/server.c',
+		     cflags='-Dmain=smbd_main')
+
+bld.SAMBA3_SUBSYSTEM('NMBD_MULTI',
+		     source='nmbd/nmbd.c',
+		     cflags='-Dmain=nmbd_main')
+
+bld.SAMBA3_SUBSYSTEM('SMBPASSWD_MULTI',
+		     source='utils/smbpasswd.c',
+		     cflags='-Dmain=smbpasswd_main')
+
+bld.SAMBA3_SUBSYSTEM('SMBCONTROL_MULTI',
+		     source='utils/smbcontrol.c',
+		     cflags='-Dmain=smbcontrol_main')
+
+bld.SAMBA3_SUBSYSTEM('SMBTREE_MULTI',
+		     source='utils/smbtree.c',
+		     cflags='-Dmain=smbtree_main')
+
+bld.SAMBA3_SUBSYSTEM('SMBSTATUS_MULTI',
+		     source=smbstatus_source,
+		     cflags='-Dmain=smbstatus_main')
+
+bld.SAMBA3_SUBSYSTEM('SMBCQUOTAS_MULTI',
+		     source='utils/smbcquotas.c',
+		     cflags='-Dmain=smbcquotas_main')
+
+bld.SAMBA3_SUBSYSTEM('SMBCACLS_MULTI',
+		     source='utils/smbcacls.c',
+		     cflags='-Dmain=smbcacls_main')
+
+bld.SAMBA3_SUBSYSTEM('NMBLOOKUP_MULTI',
+		     source='utils/nmblookup.c',
+		     cflags='-Dmain=nmblookup_main')
+
+bld.SAMBA3_SUBSYSTEM('TESTPARM_MULTI',
+		     source='utils/testparm.c',
+		     cflags='-Dmain=testparm_main')
+
+bld.SAMBA3_BINARY('samba42_multicall',
+                 source='''multicall.c 
+		 nmbd/asyncdns.c nmbd/nmbd_become_dmb.c
+                 nmbd/nmbd_become_lmb.c nmbd/nmbd_browserdb.c
+                 nmbd/nmbd_browsesync.c nmbd/nmbd_elections.c
+                 nmbd/nmbd_incomingdgrams.c nmbd/nmbd_incomingrequests.c
+                 nmbd/nmbd_lmhosts.c nmbd/nmbd_logonnames.c nmbd/nmbd_mynames.c
+                 nmbd/nmbd_namelistdb.c nmbd/nmbd_namequery.c
+                 nmbd/nmbd_nameregister.c nmbd/nmbd_namerelease.c
+                 nmbd/nmbd_nodestatus.c nmbd/nmbd_packets.c
+                 nmbd/nmbd_processlogon.c nmbd/nmbd_responserecordsdb.c
+                 nmbd/nmbd_sendannounce.c nmbd/nmbd_serverlistdb.c
+                 nmbd/nmbd_subnetdb.c nmbd/nmbd_winsproxy.c nmbd/nmbd_winsserver.c
+                 nmbd/nmbd_workgroupdb.c nmbd/nmbd_synclists.c''',
+                 deps='''smbd_base SMBD_MULTI NMBD_MULTI SMBPASSWD_MULTI
+		 SMBCONTROL_MULTI SMBTREE_MULTI SMBSTATUS_MULTI
+		 SMBCQUOTAS_MULTI SMBCACLS_MULTI NMBLOOKUP_MULTI
+		 LOCKING PROFILE PASSWD_UTIL PASSCHANGE TESTPARM_MULTI EPMD LSASD
+		 talloc param pdb''',
+                 install_path='${SBINDIR}')
+
 ########################## INCLUDES #################################
 
 bld.RECURSE('auth')
