--- a/Source/KeyPairGenerator/CMakeLists.txt
+++ b/Source/KeyPairGenerator/CMakeLists.txt
@@ -33,9 +33,7 @@ set(GENERATOR_SRC_DIR
 
 # Compiling and linking flags
 add_definitions(-std=c++11 -fpic)
-IF(ENABLE_LIBSODIUM)
-	add_definitions(-DENABLE_LIBSODIUM)
-ENDIF(ENABLE_LIBSODIUM)
+add_definitions(-DENABLE_LIBSODIUM)
 
 # Executable file name
 add_executable(KeyPairGenerator ${GENERATOR_SRC_DIR})
--- a/Source/Pcap_DNSProxy/CMakeLists.txt
+++ b/Source/Pcap_DNSProxy/CMakeLists.txt
@@ -60,12 +60,8 @@ set(MAIN_SRC_FILES
 
 # Compiling and linking flags
 add_definitions(-std=c++11 -fpic)
-IF(ENABLE_PCAP)
-	add_definitions(-DENABLE_PCAP)
-ENDIF(ENABLE_PCAP)
-IF(ENABLE_LIBSODIUM)
-	add_definitions(-DENABLE_LIBSODIUM)
-ENDIF(ENABLE_LIBSODIUM)
+add_definitions(-DENABLE_PCAP)
+add_definitions(-DENABLE_LIBSODIUM)
 
 # Executable file name
 add_executable(Pcap_DNSProxy ${MAIN_SRC_FILES})
