--- a/Source/Pcap_DNSProxy/PacketData.cpp
+++ b/Source/Pcap_DNSProxy/PacketData.cpp
@@ -335,6 +335,8 @@ size_t __fastcall AddEDNS_LabelToAdditio
 
 //Add a new EDNS/OPT Additional Resource Records.
 	auto DNS_Header = (pdns_hdr)Buffer;
+	if (DNS_Header->Additional > 0)
+		return Length;
 	DNS_Header->Additional = htons(ntohs(DNS_Header->Additional) + 1U);
 	auto DNS_Record_OPT = (pdns_record_opt)(Buffer + DataLength);
 	DNS_Record_OPT->Type = htons(DNS_RECORD_OPT);
