--- a/Source/KeyPairGenerator/CMakeLists.txt
+++ b/Source/KeyPairGenerator/CMakeLists.txt
@@ -37,6 +37,7 @@ add_definitions(-DENABLE_LIBSODIUM)
 
 # Executable file name
 add_executable(KeyPairGenerator ${GENERATOR_SRC_DIR})
+INSTALL(TARGETS KeyPairGenerator RUNTIME DESTINATION sbin)
 
 # Static libraries
 IF(ENABLE_LIBSODIUM)
--- a/Source/Pcap_DNSProxy/CMakeLists.txt
+++ b/Source/Pcap_DNSProxy/CMakeLists.txt
@@ -65,6 +65,7 @@ add_definitions(-DENABLE_LIBSODIUM)
 
 # Executable file name
 add_executable(Pcap_DNSProxy ${MAIN_SRC_FILES})
+INSTALL(TARGETS Pcap_DNSProxy RUNTIME DESTINATION sbin)
 
 # Static libraries
 target_link_libraries(Pcap_DNSProxy pthread)
