--- a/Source/KeyPairGenerator/CMakeLists.txt
+++ b/Source/KeyPairGenerator/CMakeLists.txt
@@ -39,7 +39,5 @@ add_definitions(-DENABLE_LIBSODIUM)
 add_executable(KeyPairGenerator ${GENERATOR_SRC_DIR})
 INSTALL(TARGETS KeyPairGenerator RUNTIME DESTINATION sbin)
 
-# Static libraries
-IF(ENABLE_LIBSODIUM)
-	target_link_libraries(KeyPairGenerator sodium)
-ENDIF(ENABLE_LIBSODIUM)
+# Static libraries    NO static libs plz
+# target_link_libraries(KeyPairGenerator sodium)
--- a/Source/Pcap_DNSProxy/CMakeLists.txt
+++ b/Source/Pcap_DNSProxy/CMakeLists.txt
@@ -67,11 +67,7 @@ add_definitions(-DENABLE_LIBSODIUM)
 add_executable(Pcap_DNSProxy ${MAIN_SRC_FILES})
 INSTALL(TARGETS Pcap_DNSProxy RUNTIME DESTINATION sbin)
 
-# Static libraries
-target_link_libraries(Pcap_DNSProxy pthread)
-IF(ENABLE_PCAP)
-	target_link_libraries(Pcap_DNSProxy pcap)
-ENDIF(ENABLE_PCAP)
-IF(ENABLE_LIBSODIUM)
-	target_link_libraries(Pcap_DNSProxy sodium)
-ENDIF(ENABLE_LIBSODIUM)
+# Static libraries  NO static libs plz
+# target_link_libraries(Pcap_DNSProxy pthread)
+# target_link_libraries(Pcap_DNSProxy pcap)
+# target_link_libraries(Pcap_DNSProxy sodium)
