--- a/Source/KeyPairGenerator/CMakeLists.txt
+++ b/Source/KeyPairGenerator/CMakeLists.txt
@@ -18,7 +18,7 @@
 
 
 # CMake minimum version
-cmake_minimum_required(VERSION 3.0.2)
+cmake_minimum_required(VERSION 2.8.12.2)
 
 # Project name
 project(KeyPairGenerator)
--- a/Source/Pcap_DNSProxy/CMakeLists.txt
+++ b/Source/Pcap_DNSProxy/CMakeLists.txt
@@ -18,7 +18,7 @@
 
 
 # CMake minimum version
-cmake_minimum_required(VERSION 3.0.2)
+cmake_minimum_required(VERSION 2.8.12.2)
 
 # Project name
 project(Pcap_DNSProxy)
