--- a/sim/common/ChangeLog
+++ b/sim/common/ChangeLog
@@ -1,3 +1,7 @@
+2015-02-13  Chen Gang  <gang.chen.5i5j <at> gmail.com>
+
+	*  gentmap.c: Include "string.h".
+
 2014-03-10  Mike Frysinger  <vapier@gentoo.org>
 
 	* sim-command.c (sim_do_command): Add const to cmd.
--- a/sim/common/gentmap.c
+++ b/sim/common/gentmap.c
@@ -2,6 +2,7 @@
 
 #include <stdio.h>
 #include <stdlib.h>
+#include <string.h>
 
 struct tdefs {
   char *symbol;
