Index: mt7620-wifi2716_all_dpa_20130426/rt2860v2/os/linux/rt_linux.c
===================================================================
--- mt7620-wifi2716_all_dpa_20130426.orig/rt2860v2/os/linux/rt_linux.c
+++ mt7620-wifi2716_all_dpa_20130426/rt2860v2/os/linux/rt_linux.c
@@ -38,8 +38,8 @@
 #include "dot11i_wpa.h"
 
 #if defined(CONFIG_RA_HW_NAT) || defined(CONFIG_RA_HW_NAT_MODULE)
-#include "../../../../../../net/nat/hw_nat/ra_nat.h"
-#include "../../../../../../net/nat/hw_nat/frame_engine.h"
+#include "../../../linux-2.6.36/net/nat/hw_nat/ra_nat.h"
+#include "../../../linux-2.6.36/net/nat/hw_nat/frame_engine.h"
 #endif
 
 /* TODO */
Index: mt7620-wifi2716_all_dpa_20130426/rt2860v2/os/linux/rt_profile.c
===================================================================
--- mt7620-wifi2716_all_dpa_20130426.orig/rt2860v2/os/linux/rt_profile.c
+++ mt7620-wifi2716_all_dpa_20130426/rt2860v2/os/linux/rt_profile.c
@@ -27,8 +27,8 @@
 
 #ifdef RTMP_RBUS_SUPPORT
 #if defined (CONFIG_RA_HW_NAT)  || defined (CONFIG_RA_HW_NAT_MODULE)
-#include "../../../../../../net/nat/hw_nat/ra_nat.h"
-#include "../../../../../../net/nat/hw_nat/frame_engine.h"
+#include "../../../linux-2.6.36/net/nat/hw_nat/ra_nat.h"
+#include "../../../linux-2.6.36/net/nat/hw_nat/frame_engine.h"
 #endif
 #endif /* RTMP_RBUS_SUPPORT */
 
