Index: mt7620_wifi2716_all_dpa_20130426/rt2860v2/ap/ap_cfg.c
===================================================================
--- mt7620_wifi2716_all_dpa_20130426.orig/rt2860v2/ap/ap_cfg.c
+++ mt7620_wifi2716_all_dpa_20130426/rt2860v2/ap/ap_cfg.c
@@ -13661,9 +13661,18 @@ INT Set_DyncVgaEnable_Proc(
 		RTMP_BBP_IO_WRITE8_BY_REG_ID(pAd, BBP_R195, 0x86);
 		RTMP_BBP_IO_WRITE8_BY_REG_ID(pAd, BBP_R196, 0x70);
 		RTMP_BBP_IO_WRITE8_BY_REG_ID(pAd, BBP_R195, 0x9c);
+#if 0  //noise environment
 		RTMP_BBP_IO_WRITE8_BY_REG_ID(pAd, BBP_R196, 0x27);
 		RTMP_BBP_IO_WRITE8_BY_REG_ID(pAd, BBP_R195, 0x9d);
 		RTMP_BBP_IO_WRITE8_BY_REG_ID(pAd, BBP_R196, 0x27);
+#else
+		RTMP_BBP_IO_WRITE8_BY_REG_ID(pAd, BBP_R196, 0x3D);
+		RTMP_BBP_IO_WRITE8_BY_REG_ID(pAd, BBP_R195, 0x9d);
+		if (pAd->CommonCfg.BBPCurrentBW == BW_20)
+			RTMP_BBP_IO_WRITE8_BY_REG_ID(pAd, BBP_R196, 0x40);
+		else
+			RTMP_BBP_IO_WRITE8_BY_REG_ID(pAd, BBP_R196, 0x2F);
+#endif
 	}
 	else
 	{
Index: mt7620_wifi2716_all_dpa_20130426/rt2860v2/chips/rt6352.c
===================================================================
--- mt7620_wifi2716_all_dpa_20130426.orig/rt2860v2/chips/rt6352.c
+++ mt7620_wifi2716_all_dpa_20130426/rt2860v2/chips/rt6352.c
@@ -469,11 +469,11 @@ REG_PAIR RT6352_BBP_GLRT[] = 
 	{BBP_R170, 0x30},
 	{BBP_R171, 0x30},
 	{BBP_R172, 0x0E},
-	{BBP_R173, 0x0D},
+	{BBP_R173, 0x28}, //620 patch
 	{BBP_R174, 0x28},
 	{BBP_R175, 0x21},
 	{BBP_R176, 0x1C},
-	{BBP_R177, 0x16},
+	{BBP_R177, 0x20}, //620 patch
 	{BBP_R178, 0x50},
 	{BBP_R179, 0x4A},
 	{BBP_R180, 0x43},
Index: mt7620_wifi2716_all_dpa_20130426/rt2860v2/common/rtmp_init_inf.c
===================================================================
--- mt7620_wifi2716_all_dpa_20130426.orig/rt2860v2/common/rtmp_init_inf.c
+++ mt7620_wifi2716_all_dpa_20130426/rt2860v2/common/rtmp_init_inf.c
@@ -626,6 +626,7 @@ int rt28xx_init(
 #ifdef DYNAMIC_VGA_SUPPORT
 			if (pAd->CommonCfg.MO_Cfg.bDyncVGAEnable)
 			{
+#if 0 //noise environment
 				RTMP_BBP_IO_WRITE8_BY_REG_ID(pAd, BBP_R195, 0x83);
 				RTMP_BBP_IO_WRITE8_BY_REG_ID(pAd, BBP_R196, 0x70);
 				RTMP_BBP_IO_WRITE8_BY_REG_ID(pAd, BBP_R195, 0x86);
@@ -634,6 +635,22 @@ int rt28xx_init(
 				RTMP_BBP_IO_WRITE8_BY_REG_ID(pAd, BBP_R196, 0x27);
 				RTMP_BBP_IO_WRITE8_BY_REG_ID(pAd, BBP_R195, 0x9d);
 				RTMP_BBP_IO_WRITE8_BY_REG_ID(pAd, BBP_R196, 0x27);
+#else
+				RTMP_BBP_IO_WRITE8_BY_REG_ID(pAd, BBP_R195, 0x83);
+				RTMP_BBP_IO_WRITE8_BY_REG_ID(pAd, BBP_R196, 0x70);
+				RTMP_BBP_IO_WRITE8_BY_REG_ID(pAd, BBP_R195, 0x86);
+				RTMP_BBP_IO_WRITE8_BY_REG_ID(pAd, BBP_R196, 0x70);
+				RTMP_BBP_IO_WRITE8_BY_REG_ID(pAd, BBP_R195, 0x9c);
+				//                 RTMP_BBP_IO_WRITE8_BY_REG_ID(pAd, BBP_R196, 0x27);
+				RTMP_BBP_IO_WRITE8_BY_REG_ID(pAd, BBP_R196, 0x3D);
+				RTMP_BBP_IO_WRITE8_BY_REG_ID(pAd, BBP_R195, 0x9d);
+				//                 RTMP_BBP_IO_WRITE8_BY_REG_ID(pAd, BBP_R196, 0x27);
+				if (pAd->CommonCfg.BBPCurrentBW == BW_20)
+					RTMP_BBP_IO_WRITE8_BY_REG_ID(pAd, BBP_R196, 0x40);
+				else
+					RTMP_BBP_IO_WRITE8_BY_REG_ID(pAd, BBP_R196, 0x2F);
+#endif
+
 			}
 #endif /* DYNAMIC_VGA_SUPPORT */
 		}
Index: mt7620_wifi2716_all_dpa_20130426/rt2860v2/rate_ctrl/alg_legacy.c
===================================================================
--- mt7620_wifi2716_all_dpa_20130426.orig/rt2860v2/rate_ctrl/alg_legacy.c
+++ mt7620_wifi2716_all_dpa_20130426/rt2860v2/rate_ctrl/alg_legacy.c
@@ -323,8 +323,13 @@ VOID APMlmeDynamicTxRateSwitching(
 		*/
 		if ((Rssi > -65) && (pCurrTxRate->Mode >= MODE_HTMIX))
 		{
+		#if 0
 			TrainUp		= (pCurrTxRate->TrainUp + (pCurrTxRate->TrainUp >> 1));
 			TrainDown	= (pCurrTxRate->TrainDown + (pCurrTxRate->TrainDown >> 1));
+		#else
+			TrainUp		= (pCurrTxRate->TrainUp << 1);
+			TrainDown	= (pCurrTxRate->TrainDown << 1);
+		#endif
 		}
 		else
 #endif /* DOT11_N_SUPPORT */
@@ -1702,10 +1707,26 @@ VOID MlmeOldRateAdapt(
 #endif /* TXBF_SUPPORT */
 		if (CurrRateIdx != DownRateIdx)
 		{
+
+			//620 patch
+			UCHAR *pTable = pEntry->pTable;
+			CHAR Rssi;
+			PRTMP_TX_RATE_SWITCH	pCurrTxRate;
+
+			Rssi = RTMPAvgRssi(pAd, &pEntry->RssiSample);
+			pCurrTxRate = PTX_RATE_SWITCH_ENTRY(pTable, CurrRateIdx);
+
+			if(Rssi > -60 && pCurrTxRate->CurrMCS == 3)
+			{
+				pEntry->LastSecTxRateChangeAction = RATE_NO_CHANGE;
+			}
+			else
+			{
 			pEntry->CurrTxRateIndex = DownRateIdx;
 			pEntry->LastSecTxRateChangeAction = RATE_DOWN;
 		}
 	}
+	}
 	else
 	{
 		/* Upgrade TX quality if PER <= Rate-Up threshold */
Index: mt7620_wifi2716_all_dpa_20130426/rt2860v2/rate_ctrl/ra_ctrl.c
===================================================================
--- mt7620_wifi2716_all_dpa_20130426.orig/rt2860v2/rate_ctrl/ra_ctrl.c
+++ mt7620_wifi2716_all_dpa_20130426/rt2860v2/rate_ctrl/ra_ctrl.c
@@ -183,10 +183,17 @@ UCHAR RateSwitchTable11BGN2S[] = {
     0x06, 0x21,  3, 15, 35,
     0x07, 0x21,  4, 15, 30,
     0x08, 0x20, 11, 15, 30,
+#if 0 //noise environment
     0x09, 0x20, 12, 15, 22,
     0x0a, 0x20, 13,  8, 20,
     0x0b, 0x20, 14,  8, 20,
     0x0c, 0x20, 15,  8, 20,
+#else
+    0x09, 0x20, 12, 15, 25,
+    0x0a, 0x20, 13, 12, 25,
+    0x0b, 0x20, 14, 12, 25,
+    0x0c, 0x20, 15, 12, 25,
+#endif
     0x0d, 0x22, 15,  8, 15,
 };
 
