Index: mt7620_wifi2716_all_dpa_20130426/rt2860v2/common/cmm_cfg.c
===================================================================
--- mt7620_wifi2716_all_dpa_20130426.orig/rt2860v2/common/cmm_cfg.c
+++ mt7620_wifi2716_all_dpa_20130426/rt2860v2/common/cmm_cfg.c
@@ -2240,10 +2240,10 @@ INT RTMP_COM_IoctlHandle(
 #endif /* WDS_SUPPORT */
 			{
 				HtPhyMode = pAd->ApCfg.MBSSID[pObj->ioctl_if].HTPhyMode;
-#ifdef MBSS_SUPPORT
+//#ifdef MBSS_SUPPORT
 				/* reset phy mode for MBSS */
 				MBSS_PHY_MODE_RESET(pObj->ioctl_if, HtPhyMode);
-#endif /* MBSS_SUPPORT */
+//#endif /* MBSS_SUPPORT */
 			}
 			RtmpDrvMaxRateGet(pAd, HtPhyMode.field.MODE, HtPhyMode.field.ShortGI,
 							HtPhyMode.field.BW, HtPhyMode.field.MCS,
Index: mt7620_wifi2716_all_dpa_20130426/rt2860v2/include/ap_mbss.h
===================================================================
--- mt7620_wifi2716_all_dpa_20130426.orig/rt2860v2/include/ap_mbss.h
+++ mt7620_wifi2716_all_dpa_20130426/rt2860v2/include/ap_mbss.h
@@ -43,7 +43,7 @@
 	So MBSS_PHY_MODE_RESET() can help us to adjust the correct mode &
 	maximum MCS for the BSS.
 */
-#define MBSS_PHY_MODE_RESET(__BssId, __HtPhyMode)				\
+/*#define MBSS_PHY_MODE_RESET(__BssId, __HtPhyMode)				\
 	{															\
 		UCHAR __PhyMode = pAd->ApCfg.MBSSID[__BssId].PhyMode;	\
 		if ((__PhyMode == PHY_11B) &&							\
@@ -60,7 +60,7 @@
 			__HtPhyMode.field.MCS = 7;							\
 		}														\
 	}
-
+*/
 
 /* Public function list */
 INT	Show_MbssInfo_Display_Proc(
Index: mt7620_wifi2716_all_dpa_20130426/rt2860v2/include/rtmp_comm.h
===================================================================
--- mt7620_wifi2716_all_dpa_20130426.orig/rt2860v2/include/rtmp_comm.h
+++ mt7620_wifi2716_all_dpa_20130426/rt2860v2/include/rtmp_comm.h
@@ -478,6 +478,23 @@ typedef struct _ETHEREAL_RADIO {
 #define WIRESHARK_11N_FLAG_BW40             0x80
 #endif /* MONITOR_FLAG_11N_SNIFFER_SUPPORT */
     
+#define MBSS_PHY_MODE_RESET(__BssId, __HtPhyMode)                             \
+        {                                                                                                                       \
+                UCHAR __PhyMode = pAd->ApCfg.MBSSID[__BssId].PhyMode;   \
+                if ((__PhyMode == PHY_11B) &&                                                   \
+                        (__HtPhyMode.field.MODE != MODE_CCK))                           \
+                {                                                                                                               \
+                        __HtPhyMode.field.MODE = MODE_CCK;                                      \
+                        __HtPhyMode.field.MCS = 3;                                                      \
+                }                                                                                                               \
+                else if ((__PhyMode <= PHY_11G) &&                                              \
+                                (__PhyMode != PHY_11B) &&                                               \
+                                (__HtPhyMode.field.MODE != MODE_OFDM))                  \
+                {                                                                                                               \
+                        __HtPhyMode.field.MODE = MODE_OFDM;                                     \
+                        __HtPhyMode.field.MCS = 7;                                                      \
+                }                                                                                                               \
+        }
 
 
 #endif /* __RT_COMM_H__ */
