Index: mt7620_wifi2716_all_dpa_20130426/rt2860v2/common/cmm_info.c
===================================================================
--- mt7620_wifi2716_all_dpa_20130426.orig/rt2860v2/common/cmm_info.c	2014-03-28 11:32:56.898533079 +0800
+++ mt7620_wifi2716_all_dpa_20130426/rt2860v2/common/cmm_info.c	2014-03-28 11:34:53.830533065 +0800
@@ -3668,7 +3668,9 @@
     INT         max_len = LINE_LEN;		
 	PBSS_ENTRY	pBss;
 	UINT32 TotalLen, BufLen = IW_SCAN_MAX_DATA;
+	BSS_TABLE *pScanTab;
 
+	pScanTab = &pAdapter->ScanTab;
 #ifdef CONFIG_STA_SUPPORT
 #ifdef WSC_STA_SUPPORT
 	max_len += WPS_LINE_LEN;
@@ -3705,6 +3707,8 @@
 	while ((ScanRunning(pAdapter) == TRUE) && (WaitCnt++ < 200))
 		OS_WAIT(500);	
 
+	pScanTab = &pAdapter->ScanTab;
+
 	for(i=0; i<pAdapter->ScanTab.BssNr ;i++)
 	{
 		pBss = &pAdapter->ScanTab.BssEntry[i];
