Index: mt7620_wifi2716_all_dpa_20130426/rt2860v2/common/ba_action.c
===================================================================
--- mt7620_wifi2716_all_dpa_20130426.orig/rt2860v2/common/ba_action.c
+++ mt7620_wifi2716_all_dpa_20130426/rt2860v2/common/ba_action.c
@@ -1539,9 +1539,21 @@ VOID PeerDelBAAction(
 	if (PeerDelBAActionSanity(pAd, Elem->Wcid, Elem->Msg, Elem->MsgLen))
 	{
 #ifdef PEER_DELBA_TX_ADAPT
-		Peer_DelBA_Tx_Adapt_Enable(pAd, &pAd->MacTab.Content[Elem->Wcid]);
+		MAC_TABLE_ENTRY *pEntry = NULL;
 #endif /* PEER_DELBA_TX_ADAPT */
+
 		pDelFrame = (PFRAME_DELBA_REQ)(&Elem->Msg[0]);
+
+#ifdef PEER_DELBA_TX_ADAPT
+		pEntry = &pAd->MacTab.Content[Elem->Wcid];
+
+		if ((pAd->CommonCfg.bBADecline == FALSE) && (pDelFrame->DelbaParm.TID == 0)
+			&& (!CLIENT_STATUS_TEST_FLAG(pEntry, fCLIENT_STATUS_RALINK_CHIPSET)))
+		{
+			Peer_DelBA_Tx_Adapt_Enable(pAd, &pAd->MacTab.Content[Elem->Wcid]);
+		}
+#endif /* PEER_DELBA_TX_ADAPT */
+
 		if (pDelFrame->DelbaParm.Initiator == ORIGINATOR)
 		{
 			DBGPRINT(RT_DEBUG_TRACE,("BA - PeerDelBAAction----> ORIGINATOR\n"));
@@ -2219,6 +2231,14 @@ VOID BaReOrderingBufferMaintain(
 }
 
 #ifdef PEER_DELBA_TX_ADAPT
+VOID Peer_DelBA_Tx_Adapt_Init(
+	IN PRTMP_ADAPTER pAd,
+	IN PMAC_TABLE_ENTRY pEntry)
+{
+	pEntry->bPeerDelBaTxAdaptEn = 0;
+	RTMPInitTimer(pAd, &pEntry->DelBA_tx_AdaptTimer, GET_TIMER_FUNCTION(PeerDelBATxAdaptTimeOut), pEntry, FALSE);
+}
+
 static VOID Peer_DelBA_Tx_Adapt_Enable(
 	IN PRTMP_ADAPTER pAd,
 	IN PMAC_TABLE_ENTRY pEntry)
@@ -2244,7 +2264,11 @@ static VOID Peer_DelBA_Tx_Adapt_Enable(
 	{	
 		USHORT RegId = 0;
 		UINT32 MacReg = 0, BitLUT;
+		BOOLEAN Cancelled;
+
 		pEntry->bPeerDelBaTxAdaptEn = 1;
+		RTMPCancelTimer(&pEntry->DelBA_tx_AdaptTimer, &Cancelled);
+
 		/* Enable Tx Mac look up table */
 		RTMP_IO_READ32(pAd, TX_FBK_LIMIT, &MacReg);
 		BitLUT = (MacReg & ((1 << 18)));
@@ -2260,6 +2284,7 @@ static VOID Peer_DelBA_Tx_Adapt_Enable(
 		}
 		RegId = 0x1C00 + (pEntry->Aid << 3);
 		RTMP_IO_WRITE32(pAd, RegId, 0x4007); /* Legacy OFDM / no STBC / LGI / BW20 / MCS 7 */
+		RTMPSetTimer(&pEntry->DelBA_tx_AdaptTimer, 1000); /* 1000ms */
 		DBGPRINT(RT_DEBUG_TRACE,
 				("%s():MacReg = 0x%08x, bPeerDelBaTxAdaptEn = 0x%x\n",
 				__FUNCTION__, MacReg, pEntry->bPeerDelBaTxAdaptEn));
@@ -2292,6 +2317,8 @@ static VOID Peer_DelBA_Tx_Adapt_Disable(
 	if (pEntry && pEntry->bPeerDelBaTxAdaptEn)
 	{
 		UINT32 BitLUT;
+		BOOLEAN Cancelled;
+
 		BitLUT = (pEntry->bPeerDelBaTxAdaptEn & (1 << 18));
 		if (!BitLUT)
 		{
@@ -2306,12 +2333,26 @@ static VOID Peer_DelBA_Tx_Adapt_Disable(
 		}
 		/* TODO: ressume MSC rate of the MAC look up table? */
 		pEntry->bPeerDelBaTxAdaptEn = 0;
+		RTMPCancelTimer(&pEntry->DelBA_tx_AdaptTimer, &Cancelled);
 		DBGPRINT(RT_DEBUG_TRACE,
 				("%s():bPeerDelBaTxAdaptEn = 0x%x\n",
 				__FUNCTION__, pEntry->bPeerDelBaTxAdaptEn));
 	}
 #endif /* MCS_LUT_SUPPORT */
 }
+VOID PeerDelBATxAdaptTimeOut(
+	IN PVOID SystemSpecific1,
+	IN PVOID FunctionContext,
+	IN PVOID SystemSpecific2,
+	IN PVOID SystemSpecific3)
+{
+	PMAC_TABLE_ENTRY pEntry = (PMAC_TABLE_ENTRY) FunctionContext;
+
+	DBGPRINT(RT_DEBUG_TRACE, ("%s()\n", __FUNCTION__));
+
+	/* Disable Tx Mac look up table (Ressume original setting) */
+	Peer_DelBA_Tx_Adapt_Disable(pEntry->pAd, pEntry);
+}
 #endif /* PEER_DELBA_TX_ADAPT */
 
 #endif /* DOT11_N_SUPPORT */
Index: mt7620_wifi2716_all_dpa_20130426/rt2860v2/common/cmm_data.c
===================================================================
--- mt7620_wifi2716_all_dpa_20130426.orig/rt2860v2/common/cmm_data.c
+++ mt7620_wifi2716_all_dpa_20130426/rt2860v2/common/cmm_data.c
@@ -2032,6 +2032,12 @@ VOID RTMPWriteTxWI_Data(
 #endif /* INF_AMAZON_SE */	
 
 #ifdef MCS_LUT_SUPPORT
+	if (RTMP_GET_PACKET_LOWRATE(pTxBlk->pPacket) ||
+		pTxBlk->TxFrameType == TX_MCAST_FRAME)
+	{
+		pTxWI->TXLUT = 0;
+	}
+	else
 	pTxWI->TXLUT = pAd->bUseHwTxLURate;
 #endif /* MCS_LUT_SUPPORT */
 
@@ -2195,6 +2201,12 @@ VOID RTMPWriteTxWI_Cache(
 	pTxWI->MPDUtotalByteCount = pTxBlk->MpduHeaderLen + pTxBlk->SrcBufLen;
 
 #ifdef MCS_LUT_SUPPORT
+	if (RTMP_GET_PACKET_LOWRATE(pTxBlk->pPacket) ||
+		pTxBlk->TxFrameType == TX_MCAST_FRAME)
+	{
+		pTxWI->TXLUT = 0;
+	}
+	else
 	pTxWI->TXLUT = pAd->bUseHwTxLURate;
 #endif /* MCS_LUT_SUPPORT */
 	
@@ -3013,6 +3025,9 @@ MAC_TABLE_ENTRY *MacTableInsertEntry(
 			/* Add this entry into ASIC RX WCID search table */
 			RTMP_STA_ENTRY_ADD(pAd, pEntry);
 
+#ifdef PEER_DELBA_TX_ADAPT
+			Peer_DelBA_Tx_Adapt_Init(pAd, pEntry);
+#endif /* PEER_DELBA_TX_ADAPT */
 
 #ifdef CONFIG_AP_SUPPORT
 #ifdef P2P_SUPPORT
Index: mt7620_wifi2716_all_dpa_20130426/rt2860v2/common/rtmp_timer.c
===================================================================
--- mt7620_wifi2716_all_dpa_20130426.orig/rt2860v2/common/rtmp_timer.c
+++ mt7620_wifi2716_all_dpa_20130426/rt2860v2/common/rtmp_timer.c
@@ -147,6 +147,10 @@ BUILD_TIMER_FUNCTION(P2pCliReConnectTime
 BUILD_TIMER_FUNCTION(ATEPeriodicExec);
 #endif /* RALINK_ATE */
 
+#ifdef PEER_DELBA_TX_ADAPT
+BUILD_TIMER_FUNCTION(PeerDelBATxAdaptTimeOut);
+#endif
+
 #ifdef RTMP_TIMER_TASK_SUPPORT
 static void RtmpTimerQHandle(RTMP_ADAPTER *pAd)
 {
Index: mt7620_wifi2716_all_dpa_20130426/rt2860v2/include/rtmp.h
===================================================================
--- mt7620_wifi2716_all_dpa_20130426.orig/rt2860v2/include/rtmp.h
+++ mt7620_wifi2716_all_dpa_20130426/rt2860v2/include/rtmp.h
@@ -2523,6 +2523,7 @@ typedef struct _MAC_TABLE_ENTRY {
 
 #ifdef PEER_DELBA_TX_ADAPT
 	BOOLEAN bPeerDelBaTxAdaptEn;
+	RALINK_TIMER_STRUCT DelBA_tx_AdaptTimer;
 #endif /* PEER_DELBA_TX_ADAPT */
 
 #ifdef MFB_SUPPORT
@@ -9999,5 +10000,16 @@ INT RtmpIoctl_rt_ioctl_giwscan(
 int rt_ioctl_giwscan(struct net_device *dev, struct iw_request_info *info,
                         struct iw_point *data, char *extra);
 
+#ifdef PEER_DELBA_TX_ADAPT
+VOID Peer_DelBA_Tx_Adapt_Init(
+	IN PRTMP_ADAPTER pAd,
+	IN PMAC_TABLE_ENTRY pEntry);
+
+VOID PeerDelBATxAdaptTimeOut(
+	IN PVOID SystemSpecific1,
+	IN PVOID FunctionContext,
+	IN PVOID SystemSpecific2,
+	IN PVOID SystemSpecific3);
+#endif /* PEER_DELBA_TX_ADAPT */
 #endif  /* __RTMP_H__ */
 
Index: mt7620_wifi2716_all_dpa_20130426/rt2860v2/include/rtmp_timer.h
===================================================================
--- mt7620_wifi2716_all_dpa_20130426.orig/rt2860v2/include/rtmp_timer.h
+++ mt7620_wifi2716_all_dpa_20130426/rt2860v2/include/rtmp_timer.h
@@ -214,5 +214,7 @@ DECLARE_TIMER_FUNCTION(P2pCliReConnectTi
 #ifdef RALINK_ATE
 DECLARE_TIMER_FUNCTION(ATEPeriodicExec);
 #endif /* RALINK_ATE */
-
+#ifdef PEER_DELBA_TX_ADAPT
+DECLARE_TIMER_FUNCTION(PeerDelBATxAdaptTimeOut);
+#endif /* PEER_DELBA_TX_ADAPT */
 #endif /* __RTMP_TIMER_H__ */
