Index: mt7620_wifi2716_all_dpa_20130426/rt2860v2/ap/ap.c
===================================================================
--- mt7620_wifi2716_all_dpa_20130426.orig/rt2860v2/ap/ap.c
+++ mt7620_wifi2716_all_dpa_20130426/rt2860v2/ap/ap.c
@@ -1295,9 +1295,9 @@ VOID MacTableMaintenance(
 			pEntry->PsQIdleCount ++;  
 			if (pEntry->PsQIdleCount > 2) 
 			{
-				NdisAcquireSpinLock(&pAd->irq_lock);
+				RTMP_IRQ_LOCK(&pAd->irq_lock, IrqFlags);
 				APCleanupPsQueue(pAd, &pEntry->PsQueue);
-				NdisReleaseSpinLock(&pAd->irq_lock);
+				RTMP_IRQ_UNLOCK(&pAd->irq_lock, IrqFlags);
 				pEntry->PsQIdleCount = 0;
 				WLAN_MR_TIM_BIT_CLEAR(pAd, pEntry->apidx, pEntry->Aid);
 			}
