Index: mt7620_wifi2716_all_dpa_20130426/rt2860v2/ap/ap_cfg.c
===================================================================
--- mt7620_wifi2716_all_dpa_20130426.orig/rt2860v2/ap/ap_cfg.c
+++ mt7620_wifi2716_all_dpa_20130426/rt2860v2/ap/ap_cfg.c
@@ -9158,6 +9158,7 @@ VOID RTMPIoctlStatistics(
     }
     sprintf(msg+strlen(msg), "\n");
 #endif /* APCLI_SUPPORT */
+	sprintf(msg+strlen(msg), "WscLatestMacAddr=%02x:%02x:%02x:%02x:%02x:%02x\n", PRINT_MAC(pAd->CommonCfg.LatestWSCMacAddr));
 #endif /* WSC_AP_SUPPORT */
 #ifdef RTMP_EFUSE_SUPPORT
 	if (pAd->bUseEfuse == FALSE && pAd->bFroceEEPROMBuffer == FALSE)
@@ -13038,7 +13039,7 @@ INT RTMP_AP_IoctlHandle(
 			break;
 		case CMD_RTPRIV_IOCTL_AP_SIOCGIWTXPOW:
 		{
-			UINT tmp = pAd->ApCfg.RssiSample.LastRssi0 - pAd->BbpRssiToDbmDelta;
+			UINT tmp = pAd->CommonCfg.TxPowerShow;
 			NdisCopyMemory(pData, &tmp, 4);
 			//(UINT)*(pData) =  pAd->ApCfg.RssiSample.LastRssi0 - pAd->BbpRssiToDbmDelta;
 			break;
Index: mt7620_wifi2716_all_dpa_20130426/rt2860v2/chips/rt6352.c
===================================================================
--- mt7620_wifi2716_all_dpa_20130426.orig/rt2860v2/chips/rt6352.c
+++ mt7620_wifi2716_all_dpa_20130426/rt2860v2/chips/rt6352.c
@@ -4549,7 +4549,7 @@ VOID RT6352_AsicAdjustTxPower(
 	CHAR		DeltaPowerByBbpR1 = 0; 
 	CHAR		TotalDeltaPower = 0; /* (non-positive number) including the transmit power controlled by the MAC and the BBP R1 */
 	CONFIGURATION_OF_TX_POWER_CONTROL_OVER_MAC CfgOfTxPwrCtrlOverMAC = {0};	
-
+	ULONG	ExtraPwrOverMAC = 0;
 
 #ifdef CONFIG_STA_SUPPORT
 	if (RTMP_TEST_FLAG(pAd, fRTMP_ADAPTER_IDLE_RADIO_OFF))
@@ -4720,6 +4720,9 @@ VOID RT6352_AsicAdjustTxPower(
 			DBGPRINT(RT_DEBUG_TRACE, ("TotalDeltaPower=%d, Mac 0x13B4 is 0x%08x\n", TotalDeltaPower, MacValue));
 		}
 
+		RTMP_IO_READ32(pAd, 0x1318, &ExtraPwrOverMAC);
+		pAd->CommonCfg.TxPowerShow = 16  + (TotalDeltaPower/2)+((ExtraPwrOverMAC & 0x0000FF00) >> 8);
+
 		/* Extra set MAC registers to compensate Tx power if any */
 		RTMP_CHIP_ASIC_EXTRA_POWER_OVER_MAC(pAd);
 	}
Index: mt7620_wifi2716_all_dpa_20130426/rt2860v2/common/cmm_asic.c
===================================================================
--- mt7620_wifi2716_all_dpa_20130426.orig/rt2860v2/common/cmm_asic.c
+++ mt7620_wifi2716_all_dpa_20130426/rt2860v2/common/cmm_asic.c
@@ -1728,7 +1728,7 @@ VOID AsicAdjustTxPower(
 	RTMP_CHIP_ASIC_AUTO_AGC_OFFSET_GET(
 		pAd, &DeltaPwr, &TotalDeltaPower, &TxAgcCompensate, &DeltaPowerByBbpR1);
 
-	DBGPRINT(RT_DEBUG_INFO, ("%s: DeltaPwr=%d, TotalDeltaPower=%d, TxAgcCompensate=%d, DeltaPowerByBbpR1=%d\n",
+	DBGPRINT(RT_DEBUG_TRACE, ("%s: DeltaPwr=%d, TotalDeltaPower=%d, TxAgcCompensate=%d, DeltaPowerByBbpR1=%d\n",
 			__FUNCTION__,
 			DeltaPwr,
 			TotalDeltaPower,
@@ -1759,6 +1759,7 @@ VOID AsicAdjustTxPower(
 	/* The transmit power controlled by the MAC */
 	TotalDeltaPower += DeltaPwr; 	
 
+	printk("TotalDeltaPower = %d\n",TotalDeltaPower);
 #ifdef SINGLE_SKU
 	if (pAd->CommonCfg.bSKUMode == TRUE)
 	{
Index: mt7620_wifi2716_all_dpa_20130426/rt2860v2/common/cmm_cfg.c
===================================================================
--- mt7620_wifi2716_all_dpa_20130426.orig/rt2860v2/common/cmm_cfg.c
+++ mt7620_wifi2716_all_dpa_20130426/rt2860v2/common/cmm_cfg.c
@@ -2047,7 +2047,7 @@ INT RTMP_COM_IoctlHandle(
 			PMAC_TABLE_ENTRY pMacEntry = NULL;
 #endif /* CONFIG_AP_SUPPORT */
 			RT_CMD_IW_STATS *pStats = (RT_CMD_IW_STATS *)pData;
-
+			printk("-->ra0 CMD_RTPRIV_IOCTL_INF_IW_STATUS_GET\n");
 			pStats->qual = 0;
 			pStats->level = 0;
 			pStats->noise = 0;
@@ -2124,10 +2124,19 @@ INT RTMP_COM_IoctlHandle(
 			if (CurOpMode == OPMODE_AP)
 			{
 				if (pMacEntry != NULL)
+				{
 					pStats->level =
 						RTMPMaxRssi(pAd, pMacEntry->RssiSample.AvgRssi0,
 										pMacEntry->RssiSample.AvgRssi1,
 										pMacEntry->RssiSample.AvgRssi2);
+				}
+				else
+				{
+					if ((pStats->priv_flags != INT_APCLI))
+						pStats->level =RTMPMaxRssi(pAd, pAd->ApCfg.RssiSample.AvgRssi0,
+												pAd->ApCfg.RssiSample.AvgRssi1,
+												pAd->ApCfg.RssiSample.AvgRssi2);
+				}
 #ifdef P2P_APCLI_SUPPORT
 				else
 					pStats->level =
@@ -2139,10 +2148,6 @@ INT RTMP_COM_IoctlHandle(
 #endif /* CONFIG_AP_SUPPORT */
 
 #ifdef CONFIG_AP_SUPPORT
-			pStats->level = RTMPMaxRssi(pAd, pAd->ApCfg.RssiSample.AvgRssi0,
-										pAd->ApCfg.RssiSample.AvgRssi1,
-										pAd->ApCfg.RssiSample.AvgRssi2);
-
 			pStats->noise = RTMPMaxRssi(pAd, pAd->ApCfg.RssiSample.AvgRssi0,
 										pAd->ApCfg.RssiSample.AvgRssi1,
 										pAd->ApCfg.RssiSample.AvgRssi2) -
Index: mt7620_wifi2716_all_dpa_20130426/rt2860v2/common/wsc.c
===================================================================
--- mt7620_wifi2716_all_dpa_20130426.orig/rt2860v2/common/wsc.c
+++ mt7620_wifi2716_all_dpa_20130426/rt2860v2/common/wsc.c
@@ -7053,6 +7053,7 @@ VOID WscInit(
 				APUpdateBeaconFrame(pAd, pWscControl->EntryIfIdx & 0x0F);
 			}
 		}
+		NdisZeroMemory(pAd->CommonCfg.LatestWSCMacAddr,MAC_ADDR_LEN);
 #endif /* CONFIG_AP_SUPPORT */        
 	}
 
Index: mt7620_wifi2716_all_dpa_20130426/rt2860v2/include/rtmp.h
===================================================================
--- mt7620_wifi2716_all_dpa_20130426.orig/rt2860v2/include/rtmp.h
+++ mt7620_wifi2716_all_dpa_20130426/rt2860v2/include/rtmp.h
@@ -1895,6 +1895,7 @@ typedef struct _COMMON_CONFIG {
 	MO_CFG_STRUCT MO_Cfg;	/* data structure for mitigating microwave interference */
 #endif /* defined(MICROWAVE_OVEN_SUPPORT) || defined(DYNAMIC_VGA_SUPPORT) */
 	BOOLEAN bEnTemperatureTrack;
+	UCHAR TxPowerShow;
 } COMMON_CONFIG, *PCOMMON_CONFIG;
 
 #ifdef DBG_CTRL_SUPPORT
