--- a/arch/mips/ath79/early_printk.c
+++ b/arch/mips/ath79/early_printk.c
@@ -114,7 +114,8 @@
 	case REV_ID_MAJOR_AR9341:
 	case REV_ID_MAJOR_AR9342:
 	case REV_ID_MAJOR_AR9344:
-	case REV_ID_MAJOR_QCA9533:
+	case REV_ID_MAJOR_QCA9533_V1:
+	case REV_ID_MAJOR_QCA9533_V2:
 	case REV_ID_MAJOR_QCA9556:
 	case REV_ID_MAJOR_QCA9558:
 	case REV_ID_MAJOR_TP9343:
--- a/arch/mips/ath79/setup.c
+++ b/arch/mips/ath79/setup.c
@@ -152,10 +152,13 @@
 		rev = id & AR934X_REV_ID_REVISION_MASK;
 		break;
 
-	case REV_ID_MAJOR_QCA9533:
+	case REV_ID_MAJOR_QCA9533_V1:
+	case REV_ID_MAJOR_QCA9533_V2:
 		ath79_soc = ATH79_SOC_QCA9533;
 		chip = "9533";
 		rev = id & QCA953X_REV_ID_REVISION_MASK;
+		if (major == REV_ID_MAJOR_QCA9533_V2)
+			rev = 2;
 		break;
 
 	case REV_ID_MAJOR_QCA9556:
--- a/arch/mips/include/asm/mach-ath79/ar71xx_regs.h
+++ b/arch/mips/include/asm/mach-ath79/ar71xx_regs.h
@@ -709,7 +709,8 @@
 #define REV_ID_MAJOR_AR9341		0x0120
 #define REV_ID_MAJOR_AR9342		0x1120
 #define REV_ID_MAJOR_AR9344		0x2120
-#define REV_ID_MAJOR_QCA9533		0x0140
+#define REV_ID_MAJOR_QCA9533_V1	0x0140
+#define REV_ID_MAJOR_QCA9533_V2	0x0160
 #define REV_ID_MAJOR_QCA9556		0x0130
 #define REV_ID_MAJOR_QCA9558		0x1130
 #define REV_ID_MAJOR_TP9343		0x0150
