--- a/Tests/RunCMake/CMakeLists.txt
+++ b/Tests/RunCMake/CMakeLists.txt
@@ -143,14 +143,6 @@ add_RunCMake_test(interface_library)
 add_RunCMake_test(no_install_prefix)
 add_RunCMake_test(configure_file)
 
-find_package(Qt4 QUIET)
-find_package(Qt5Core QUIET)
-if (QT4_FOUND AND Qt5Core_FOUND AND NOT Qt5Core_VERSION VERSION_LESS 5.1.0)
-  add_RunCMake_test(IncompatibleQt)
-endif()
-if (QT4_FOUND)
-  add_RunCMake_test(ObsoleteQtMacros -DQT_QMAKE_EXECUTABLE:FILEPATH=${QT_QMAKE_EXECUTABLE})
-endif()
 
 find_package(PkgConfig QUIET)
 if(PKG_CONFIG_FOUND)
--- a/Tests/CMakeLists.txt
+++ b/Tests/CMakeLists.txt
@@ -318,9 +318,6 @@ if(BUILD_TESTING)
 
   list(APPEND TEST_BUILD_DIRS ${CMake_TEST_INSTALL_PREFIX})
 
-  if(NOT QT4_FOUND)
-    find_package(Qt4 QUIET)
-  endif()
 
   if(QT4_FOUND)
     # test whether the Qt4 which has been found works, on some machines
@@ -1143,7 +1140,6 @@ ${CMake_BINARY_DIR}/bin/cmake -DDIR=dev
     set(run_autouic_test QtAutoUicInterface)
   endif()
 
-  find_package(Qt5Widgets QUIET NO_MODULE)
   if(Qt5Widgets_FOUND)
     add_test(Qt5Autogen ${CMAKE_CTEST_COMMAND}
       --build-and-test
